File: content/tracks/dsa-python/concepts/06-stacks/stacks_test.py
import sys
from stub import is_valid_parentheses, daily_temperatures, largest_rectangle_in_histogram

passed = 0
failed = 0


def test(name: str, condition: bool, detail: str = "") -> None:
    global passed, failed
    if condition:
        print(f"  PASS: {name}")
        passed += 1
    else:
        msg = f" — {detail}" if detail else ""
        print(f"  FAIL: {name}{msg}")
        failed += 1


# is_valid_parentheses
test("parens: simple ()", is_valid_parentheses("()") is True)
test("parens: all types", is_valid_parentheses("()[]{}")  is True)
test("parens: nested {[]}", is_valid_parentheses("{[]}") is True)
test("parens: mismatched (]", is_valid_parentheses("(]") is False)
test("parens: interleaved ([)]", is_valid_parentheses("([)]") is False)
test("parens: empty string", is_valid_parentheses("") is True)
test("parens: unclosed open", is_valid_parentheses("(") is False)
test("parens: close without open", is_valid_parentheses(")") is False)
test("parens: deeply nested ((()))", is_valid_parentheses("((()))") is True)

# daily_temperatures
r = daily_temperatures([73, 74, 75, 71, 69, 72, 76, 73])
test("temps: standard", r == [1, 1, 4, 2, 1, 1, 0, 0], f"got {r}")
r = daily_temperatures([30, 40, 50, 60])
test("temps: increasing", r == [1, 1, 1, 0], f"got {r}")
r = daily_temperatures([30, 60, 90])
test("temps: three ascending", r == [1, 1, 0], f"got {r}")
r = daily_temperatures([90, 80, 70, 60])
test("temps: decreasing", r == [0, 0, 0, 0], f"got {r}")
r = daily_temperatures([70])
test("temps: single day", r == [0], f"got {r}")
r = daily_temperatures([55, 55, 55])
test("temps: all equal", r == [0, 0, 0], f"got {r}")

# largest_rectangle_in_histogram
r = largest_rectangle_in_histogram([2, 1, 5, 6, 2, 3])
test("histogram: standard (expect 10)", r == 10, f"got {r}")
r = largest_rectangle_in_histogram([2, 4])
test("histogram: two bars (expect 4)", r == 4, f"got {r}")
r = largest_rectangle_in_histogram([1])
test("histogram: single bar (expect 1)", r == 1, f"got {r}")
r = largest_rectangle_in_histogram([6, 6, 6])
test("histogram: all equal (expect 18)", r == 18, f"got {r}")
r = largest_rectangle_in_histogram([1, 2, 3, 4, 5])
test("histogram: increasing (expect 9)", r == 9, f"got {r}")
r = largest_rectangle_in_histogram([5, 4, 3, 2, 1])
test("histogram: decreasing (expect 9)", r == 9, f"got {r}")
r = largest_rectangle_in_histogram([0, 0, 0])
test("histogram: all zeros (expect 0)", r == 0, f"got {r}")

print(f"\n{passed} passed, {failed} failed")
if failed > 0:
    sys.exit(1)

File: content/tracks/dsa-typescript/concepts/06-stacks/stacks_test.ts
import { isValidParentheses, dailyTemperatures, largestRectangleInHistogram } from "./stub";

let passed = 0;
let failed = 0;

function test(name: string, fn: () => void): void {
  try {
    fn();
    console.log(`  PASS: ${name}`);
    passed++;
  } catch (e: unknown) {
    const msg = e instanceof Error ? e.message : String(e);
    console.log(`  FAIL: ${name} — ${msg}`);
    failed++;
  }
}

function assertEqual<T>(actual: T, expected: T): void {
  const a = JSON.stringify(actual);
  const b = JSON.stringify(expected);
  if (a !== b) throw new Error(`expected ${b}, got ${a}`);
}

// --- isValidParentheses ---

test("isValidParentheses: simple parens", () => {
  assertEqual(isValidParentheses("()"), true);
});

test("isValidParentheses: all three types", () => {
  assertEqual(isValidParentheses("()[]{}"), true);
});

test("isValidParentheses: nested valid", () => {
  assertEqual(isValidParentheses("{[]}"), true);
});

test("isValidParentheses: mismatched type", () => {
  assertEqual(isValidParentheses("(]"), false);
});

test("isValidParentheses: interleaved invalid", () => {
  assertEqual(isValidParentheses("([)]"), false);
});

test("isValidParentheses: unclosed bracket", () => {
  assertEqual(isValidParentheses("("), false);
});

test("isValidParentheses: close with empty stack", () => {
  assertEqual(isValidParentheses(")"), false);
});

test("isValidParentheses: empty string", () => {
  assertEqual(isValidParentheses(""), true);
});

// --- dailyTemperatures ---

test("dailyTemperatures: standard case", () => {
  assertEqual(dailyTemperatures([73, 74, 75, 71, 69, 72, 76, 73]), [1, 1, 4, 2, 1, 1, 0, 0]);
});

test("dailyTemperatures: all decreasing", () => {
  assertEqual(dailyTemperatures([5, 4, 3, 2, 1]), [0, 0, 0, 0, 0]);
});

test("dailyTemperatures: all increasing", () => {
  assertEqual(dailyTemperatures([1, 2, 3, 4, 5]), [1, 1, 1, 1, 0]);
});

test("dailyTemperatures: single element", () => {
  assertEqual(dailyTemperatures([50]), [0]);
});

test("dailyTemperatures: two elements warmer", () => {
  assertEqual(dailyTemperatures([30, 60]), [1, 0]);
});

test("dailyTemperatures: two elements cooler", () => {
  assertEqual(dailyTemperatures([60, 30]), [0, 0]);
});

// --- largestRectangleInHistogram ---

test("largestRectangleInHistogram: standard case", () => {
  assertEqual(largestRectangleInHistogram([2, 1, 5, 6, 2, 3]), 10);
});

test("largestRectangleInHistogram: two bars", () => {
  assertEqual(largestRectangleInHistogram([2, 4]), 4);
});

test("largestRectangleInHistogram: single bar", () => {
  assertEqual(largestRectangleInHistogram([5]), 5);
});

test("largestRectangleInHistogram: uniform height", () => {
  assertEqual(largestRectangleInHistogram([3, 3, 3, 3]), 12);
});

test("largestRectangleInHistogram: pyramid shape", () => {
  assertEqual(largestRectangleInHistogram([1, 2, 3, 2, 1]), 6);
});

test("largestRectangleInHistogram: all ones", () => {
  assertEqual(largestRectangleInHistogram([1, 1, 1, 1, 1]), 5);
});

test("largestRectangleInHistogram: descending", () => {
  assertEqual(largestRectangleInHistogram([5, 4, 3, 2, 1]), 9);
});

console.log(`\n${passed} passed, ${failed} failed`);
if (failed > 0) process.exit(1);
